<img src="https://www.politecnicos.com.br/img/075.jpg" alt="Grupo Turing" height="420" width="420">

# Análise de dados - Spotify

Por: Camilla de Oliveira Fonseca

Essa é uma análise do dataset "spotify.csv" proposta como mini-projeto de treinamento do Grupo Turing.

### Descrição do dataset
retirada [desse site](https://developer.spotify.com/documentation/web-api/reference/tracks/get-audio-features/)

**Features:**

* **song_duration_ms:** _(int)_ The duration of the track in milliseconds.

* **key:** _(int)_ The estimated overall key of the track. Integers map to pitches using standard Pitch Class notation
> E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.

* **audio_mode:** _(int)_Mode indicates the modality (major or minor) of a track, the type of scale from which its melodic content is derived.
> Major is represented by 1 and minor is 0.

* **time_signature:** _(int)_ An estimated overall time signature of a track. The time signature (meter) is a notational convention to specify how many beats are in each bar (or measure).

* **acousticness:** _(float)_ A confidence measure from 0.0 to 1.0 of whether the track is acoustic.
> 1.0 represents high confidence the track is acoustic.

* **danceability:** _(float)_ Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity.
> A value of 0.0 is least danceable and 1.0 is most danceable.

* **energy:** _(float)_ Energy is a measure from *0.0 to 1.0* and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale.

* **instrumentalness:** _(float)_ Detects the presence of an audience in the recording.
> Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.

* **loudness:** _(float)_ The overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude).
> Values typical range between -60 and 0 db.

* **speechiness:** _(float)_ Speechiness detects the presence of spoken words in a track.
> The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.

* **audio_valence:** _(float)_ A measure from *0.0 to 1.0* describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).

* **tempo:** _(float)_ The overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.

* **song_popularity:** _(int)_ Song ratings of spotify audience.

* **liveness:** _(float)_ Detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live.
> A value above 0.8 provides strong likelihood that the track is live.

In [1]:
#Importando as bibliotecas necessarias
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


%matplotlib inline


In [75]:
#Leitura do dataset
df = pd.read_csv("spotify.csv", index_col = 0) #index_col pq o dataset ja tem uma coluna de index

# Explorando e preparando os dados


## Problemas de tipos de dado e consistência

In [77]:
df.head(15)

song_name song_popularity song_duration_ms  \
0   Boulevard of Broken Dreams              73           262333   
1                   In The End              66           216933   
2            Seven Nation Army              76           231733   
3                   By The Way              74           216933   
4            How You Remind Me              56           223826   
5             Bring Me To Life              80           235893   
6                  Last Resort              81           199893   
7     Are You Gonna Be My Girl              76           213800   
8               Mr. Brightside              80           222586   
9                  Sex on Fire              81           203346   
10                  The Middle              78           168253   
11                        Numb              63           185586   
12             Smooth Criminal              75           209266   
13                  Can't Stop              81           269000   
14                  Chop Suey!              69           210240   

              acousticness danceability              energy  \
0   0.005520000000000001kg   0.496mol/L               0.682   
1                 0.0103kg   0.542mol/L               0.853   
2                0.00817kg   0.737mol/L               0.463   
3                 0.0264kg   0.451mol/L                0.97   
4               0.000954kg   0.447mol/L  0.7659999999999999   
5                0.00895kg   0.316mol/L               0.945   
6               0.000504kg   0.581mol/L               0.887   
7                0.00148kg   0.613mol/L               0.953   
8                0.00108kg    0.33mol/L               0.936   
9                0.00172kg   0.542mol/L               0.905   
10                0.0424kg   0.629mol/L               0.897   
11                0.0046kg   0.496mol/L               0.863   
12               0.00434kg   0.647mol/L               0.964   
13                0.0179kg   0.618mol/L               0.938   
14              0.000353kg    0.42mol/L               0.929   

         instrumentalness   key             liveness             loudness  \
0                2.94e-05   8.0               0.0589               -4.095   
1                     0.0   3.0                0.108               -6.407   
2                   0.447   0.0                0.255  -7.8279999999999985   
3                 0.00355   0.0                0.102               -4.938   
4                     0.0  10.0                0.113               -5.065   
5                1.85e-06   4.0                0.396               -3.169   
6                 0.00111   4.0                0.268               -3.659   
7                0.000582   2.0                0.152               -3.435   
8                     0.0   1.0               0.0926                -3.66   
9                  0.0104   9.0                0.136               -5.653   
10                    0.0   2.0                0.263               -3.401   
11                    0.0   9.0                0.639  -4.1530000000000005   
12                 0.0036   9.0                 0.15               -4.225   
13                    0.0   9.0  0.16699999999999998               -3.442   
14  0.0007469999999999999   7.0                0.122               -3.899   

   audio_mode           speechiness    tempo time_signature  audio_valence  
0           1                0.0294   167.06              4          0.474  
1           0                0.0498  105.256              4          0.370  
2           1                0.0792  123.881              4          0.324  
3           1                 0.107  122.444              4          0.198  
4           1                0.0313  172.011              4          0.574  
5           0                 0.124  189.931              4          0.320  
6           0                0.0624   90.578              4          0.724  
7           1                0.0855  105.046              4          0.537  
8           1

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18835 entries, 0 to 18834
Data columns (total 15 columns):
song_name           18835 non-null object
song_popularity     18835 non-null object
song_duration_ms    18835 non-null object
acousticness        18835 non-null object
danceability        18835 non-null object
energy              18835 non-null object
instrumentalness    18835 non-null object
key                 18835 non-null float64
liveness            18835 non-null object
loudness            18835 non-null object
audio_mode          18835 non-null object
speechiness         18835 non-null object
tempo               18835 non-null object
time_signature      18835 non-null object
audio_valence       18834 non-null float64
dtypes: float64(2), object(13)
memory usage: 2.3+ MB


Vemos que a única feature que está armazenada no tipo de dado correto é *audio_valence*.

Além disso, os valores de *acousticness* e *danceability* estão salvos com "kg" e "mol/L" que, além de não fazerem sentido, precisam ser retirados para converter os dados em *float*.

Vemos também que há 1 dado nulo em _audio_valence_.

In [81]:
#Retirando kg e mol/L

df["acousticness"] = df["acousticness"].str.strip("kg")
df["danceability"] = df["danceability"].str.strip("mol/L")

#Retirando nulos
df.dropna(inplace = True)


Ao tentar converter os dados para *float*, encontrei alguns valores marcado com "nao_sei". Usar a função unique() seria uma forma de verificar se há valores inadequados como esse, porém a maioria das features tem uma quantidade muito grande de valores diferentes (uma vez que na verdade são dados contínuos) e isso não é possível.

Como o dataset é grande, provavelmente não haverá problemas em excluir as linhas com 'nao_sei'.

In [83]:
ftrs_str = ["song_name","song_popularity", "song_duration_ms", "acousticness", "danceability", "energy", "instrumentalness", "liveness", "loudness", "audio_mode", "speechiness", "tempo", "time_signature"]

for ftr in ftrs_str:
    df = df[df[ftr] != 'nao_sei']
    df[ftr] = df[ftr].str.strip("nao_sei")

In [85]:
ftrs_unq = ["key", "audio_mode", "time_signature", "song_popularity"]

for ftr in ftrs_unq:
    print(df[ftr].unique())

[ 8.  3.  0. 10.  4.  2.  1.  9.  7. 11.  5.  6.]
['1' '0']
['4' '3' '1' '5' '2800000000' '0']
['73' '66' '76' '74' '56' '80' '81' '78' '63' '75' '69' '77' '71' '62'
 '79' '13' '28' '11' '65' '70' '60' '72' '57' '64' '61' '67' '94' '98'
 '59' '87' '85' '58' '92' '83' '44' '47' '54' '49' '52' '95' '45' '38'
 '46' '53' '39' '88' '68' '37' '43' '84' '40' '41' '10' '31' '48' '24'
 '29' '51' '4' '7' '50' '42' '30' '21' '55' '14' '33' '8' '16' '34' '26'
 '15' '19' '5' '3' '22' '36' '32' '35' '9' '82' '25' '86' '12' '18' '27'
 '17' '6' '20' '0' '90' '93' '91' '89' '99' '97' '96' '1' '2' '23' '100']


Para as features que são possíveis explorar com unique(), chama a atenção um valor de *time_signature* muito alto (280000000) e inconsistente.

In [87]:
df[df["time_signature"] == '2800000000']

song_name song_popularity song_duration_ms acousticness  \
149  What Makes You Beautiful              69           198053      0.00761   

           danceability              energy instrumentalness  key liveness  \
149  0.7290000000000001  0.7709999999999999              0.0  4.0    0.087   

    loudness audio_mode speechiness    tempo time_signature  audio_valence  
149   -2.451          1      0.0725  125.011     2800000000          0.873

Pesquisando na internet, vemos que o tempo de "What Makes You Beautiful" é 4, assim podemos arrumar.

In [89]:
df.loc[149,"time_signature"] = '4'

df.loc[149]

song_name           What Makes You Beautiful
song_popularity                           69
song_duration_ms                      198053
acousticness                         0.00761
danceability              0.7290000000000001
energy                    0.7709999999999999
instrumentalness                         0.0
key                                        4
liveness                               0.087
loudness                              -2.451
audio_mode                                 1
speechiness                           0.0725
tempo                                125.011
time_signature                             4
audio_valence                          0.873
Name: 149, dtype: object

Fiquei um tanto curiosa também com o valor 0 para essa feature, pois não sabia se fazia sentido.

In [91]:
df[df["time_signature"] == '0']

song_name song_popularity  \
7119   Gina Rodriguez - Latinos Trending Intr               0   
11171                                     Aur              50   
18120                                  Imma B              61   

      song_duration_ms acousticness danceability energy instrumentalness  \
7119             12000          0.7          0.0  0.493              0.0   
11171           102536       0.0774          0.0   0.56            0.963   
18120           257560        0.184        0.619  0.539              0.0   

        key             liveness loudness audio_mode speechiness    tempo  \
7119    7.0                0.457   -6.102          1         0.0      0.0   
11171  11.0                0.589   -9.866          1         0.0      0.0   
18120   0.0  0.28800000000000003     -6.9          1       0.387  145.618   

      time_signature  audio_valence  
7119               0          0.000  
11171              0          0.000  
18120              0          0.424

Pesquisando na internet, vi que a notação "0" pode ser usada para músicas com compassos livres ou irregulares, que parece ser o caso das músicas acima.

Agora podemos converter os dados para seus tipos corretos.

As features *audio_mode* e *time_signature* são categóricas e, apesar de constarem como *int* na descrição do dataset, preferi convertê-las para *category*.

In [93]:
ftrs_float = ["acousticness", "danceability", "energy", "instrumentalness", "liveness", "loudness", "speechiness", "tempo"]

ftrs_int = ["song_popularity", "song_duration_ms", "key"]

ftrs_cat = ["audio_mode", "time_signature"]

#Corvertendo para float
for ftr in ftrs_float:
    df[ftr] = df[ftr].astype('float')

#Convertendo para int
for ftr in ftrs_int:
    df[ftr] = df[ftr].astype(np.int64)
    # quando tentei .astype('int') deu erro com a seguinte msg: Python int too large to convert to C long

#Convertendo para categorico
for ftr in ftrs_cat:
    df[ftr] = df[ftr].astype('category')

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18823 entries, 0 to 18834
Data columns (total 15 columns):
song_name           18823 non-null object
song_popularity     18823 non-null int64
song_duration_ms    18823 non-null int64
acousticness        18823 non-null float64
danceability        18823 non-null float64
energy              18823 non-null float64
instrumentalness    18823 non-null float64
key                 18823 non-null int64
liveness            18823 non-null float64
loudness            18823 non-null float64
audio_mode          18823 non-null category
speechiness         18823 non-null float64
tempo               18823 non-null float64
time_signature      18823 non-null category
audio_valence       18823 non-null float64
dtypes: category(2), float64(9), int64(3), object(1)
memory usage: 2.7+ MB


In [97]:
df.head()

song_name  song_popularity  song_duration_ms  acousticness  \
0  Boulevard of Broken Dream               73            262333      0.005520   
1                 In The End               66            216933      0.010300   
2          Seven Nation Army               76            231733      0.008170   
3                 By The Way               74            216933      0.026400   
4           How You Remind M               56            223826      0.000954   

   danceability  energy  instrumentalness  key  liveness  loudness audio_mode  \
0         0.496   0.682          0.000029    8    0.0589    -4.095          1   
1         0.542   0.853          0.000000    3    0.1080    -6.407          0   
2         0.737   0.463          0.447000    0    0.2550    -7.828          1   
3         0.451   0.970          0.003550    0    0.1020    -4.938          1   
4         0.447   0.766          0.000000   10    0.1130    -5.065          1   

   speechiness    tempo time_signature  audio_valence  
0       0.0294  167.060              4          0.474  
1       0.0498  105.256              4          0.370  
2       0.0792  123.881              4          0.324  
3       0.1070  122.444              4          0.198  
4       0.0313  172.011              4          0.574

In [99]:
df[ftrs_cat].describe()

audio_mode time_signature
count       18823          18823
unique          2              5
top             1              4
freq        11825          17744

Vemos que os valores de *time_signatue* estão **bem** concentrados em 4, a frequência ultrapassa 90%.

In [101]:
df.describe().T

count          mean           std           min  \
song_popularity   18823.0  5.299033e+01  2.190882e+01  0.000000e+00   
song_duration_ms  18823.0 -9.841578e+08  1.350534e+11 -1.852891e+13   
acousticness      18823.0  2.585292e-01  2.887133e-01  1.020000e-06   
danceability      18823.0  6.334021e-01  1.566931e-01  0.000000e+00   
energy            18823.0  6.450035e-01  2.140746e-01  1.070000e-03   
instrumentalness  18823.0  7.800799e-02  2.215674e-01  0.000000e+00   
key               18823.0  5.288902e+00  3.614212e+00  0.000000e+00   
liveness          18823.0  1.796629e-01  1.440198e-01  1.090000e-02   
loudness          18823.0 -7.447085e+00  3.826882e+00 -3.876800e+01   
speechiness       18823.0  1.020912e-01  1.043930e-01  0.000000e+00   
tempo             18823.0  1.210770e+02  2.871319e+01  0.000000e+00   
audio_valence     18823.0  5.279494e-01  2.446100e-01  0.000000e+00   

                          25%            50%           75%          max  
song_popularity       40.0000      56.000000      69.00000      100.000  
song_duration_ms  184339.5000  211306.000000  242844.00000  1799346.000  
acousticness           0.0241       0.132000       0.42400        0.996  
danceability           0.5330       0.645000       0.74800        0.987  
energy                 0.5100       0.674000       0.81500        0.999  
instrumentalness       0.0000       0.000011       0.00257        0.997  
key                    2.0000       5.000000       8.00000       11.000  
liveness               0.0929       0.122000       0.22100        0.986  
loudness              -9.0440      -6.555000      -4.90900        1.585  
speechiness            0.0378       0.055500       0.11900        0.941  
tempo                 98.4155     120.013000     139.93100      242.318  
audio_valence          0.3350       0.527000       0.72500        0.984

Observamos problemas em _song_duration_ms_, uma vez que a duração de uma música não pode ser negativa, e em _loudness_, já que seu máximo é 0.

In [103]:
df[df["song_duration_ms"] <= 0]

song_name  song_popularity  song_duration_ms  acousticness  danceability  \
62   White B               59   -18528908788662        0.0983          0.83   

    energy  instrumentalness  key  liveness  loudness audio_mode  speechiness  \
62   0.497               0.0    1    0.0906     -6.94          1        0.052   

      tempo time_signature  audio_valence  
62  127.965              4          0.634

In [105]:
df[df["loudness"] > 0]

song_name  song_popularity  \
2503            We Are Your Friends - Radio Edit               57   
2819           Search and Destroy - Iggy Pop Mix               55   
6219                                       Polar               55   
7980                    Gladiator - Instrumental               14   
8950                          Wake Up (RIOT VIP)               52   
13006       We Are Your Friends - Justice Vs Sim               58   
15623  Long Story Short - Bodybangers Remix Edit               33   

       song_duration_ms  acousticness  danceability  energy  instrumentalness  \
2503             164960       0.01020         0.646   0.977          0.769000   
2819             208133       0.00353         0.235   0.977          0.006040   
6219             236240       0.00679         0.795   0.994          0.904000   
7980             198000       0.01070         0.780   0.972          0.029600   
8950             292915       0.02510         0.341   0.995          0.000070   
13006            262773       0.01040         0.615   0.970          0.384000   
15623            206718       0.01610         0.762   0.977          0.000011   

       key  liveness  loudness audio_mode  speechiness    tempo  \
2503     9     0.201     1.585          0       0.1670  123.016   
2819     6     0.172     0.878          1       0.1070  152.952   
6219     8     0.057     0.525          0       0.0483  126.023   
7980     6     0.100     0.119          0       0.2530  137.056   
8950     7     0.713     0.052          1       0.6260  160.311   
13006    9     0.178     1.342          0       0.1190  122.993   
15623    5     0.535     0.198          0       0.0934  128.012   

      time_signature  audio_valence  
2503               4         0.6060  
2819               4         0.2410  
6219               4         0.8650  
7980               4         0.9580  
8950               4         0.0801  
13006              4         0.5070  
15623              4         0.6160

Como são poucos e seria complicado corrigir esses valores, optei por retirá-los.

In [107]:
df = df[(df["song_duration_ms"] > 0) & (df["loudness"] <= 0)]

## Dados duplicados

In [109]:
antes = len(df)
print(antes)

18815


In [111]:
# Retirando duplicados completos
df.drop_duplicates(inplace = True)

# Analisando nomes duplicados
duplicates = df.duplicated(subset = "song_name", keep = False)
df[duplicates].sort_values(by = "song_name")

song_name  song_popularity  \
5614                       tears left to cry               92   
4644                       tears left to cry               74   
2229                 (Don't Fear) The Reaper               19   
459                  (Don't Fear) The Reaper               76   
660                (I Just) Died In Your Arm               69   
2306               (I Just) Died In Your Arm               72   
807            (I've Had) The Time of My Lif               21   
2324           (I've Had) The Time of My Lif               63   
16924           (No One Knows Me) Like the P               19   
2621            (No One Knows Me) Like the P               67   
397    (You Make Me Feel Like) A Natural Wom               62   
17273  (You Make Me Feel Like) A Natural Wom               71   
18398                         1-800-273-8255                8   
10003                         1-800-273-8255               86   
4718                                     123               61   
8377                                     123               73   
6455                                    1901               68   
2506                                    1901               66   
4088                                    1942               81   
12495                                   1942               49   
12397                                   1950               66   
4663                                    1950               83   
4924                                    1979               54   
940                                     1979               19   
1111                  1979 - Remastered 2012               76   
2955                  1979 - Remastered 2012               47   
8618                                    1985               70   
8116                                    1985               61   
8358                                    2002               87   
11326                                   2002               85   
...                                      ...              ...   
4654                              Youngblood               89   
15769                                Younger               62   
13558                                Younger               76   
791                                 Your Lov               72   
4244                                Your Lov               80   
14706                               Your Lov                0   
16048                               Your Lov               49   
10116                              Your Song               64   
13966                              Your Song               80   
16971                              Your Song               56   
9586                                   Youth               65   
15659                                  Youth               70   
1219      ZEZE (feat. Travis Scott & Offset)               91   
7435      ZEZE (feat. Travis Scott & Offset)               71   
8407                                    Zomb               78   
9433                                    Zomb                1   
1092                                    Zomb               82   
1755                                    Zomb               30   
2612                                    Zomb               63   
4748                              better off               51   
4711                              better off               25   
11719                                   brok               82   
6469                                    brok               49   
9629                                    brok                5   
14817                                hometow               64   
18016                                hometow               58   
3406                                rockstar               31   
3699                                rockstar               27   
15670                    these are my friend               38   
12388                    these are my friend               65   

    

In [113]:
depois1 = len(df)
print(antes - depois1)

3903


A questão dos duplicados se tratando de música é um tanto singular. Vemos que há várias linhas com mesmo nome, porém demais valores diferentes. Isso provavelmente se deve ao fato de haver músicas distintas com o mesmo nome ou versões diferentes da mesma música. No entanto, seria difícil músicas ou versão diferentes terem todos os valores exatamente iguais, por isso exclui duplicados completos com *.drop_duplicates()*. Por enquanto, 3903 linhas duplicadas foram retiradas.

Pode haver também duplicados não excluidos com *.drop_duplicates()* devido a pequenas variações nos floats ou na duração, já que ela está em milissegundos.

In [115]:
colunas = ["song_name", "song_popularity", "key", "audio_mode", "time_signature"]
duplicates = df.duplicated(subset = colunas, keep=False)
df[duplicates].sort_values(by = "song_name")

song_name  song_popularity  song_duration_ms  \
11960  Always Remember Us This Way               79            210200   
5910   Always Remember Us This Way               79            210200   
15407                         Amig               45            279946   
4194                          Amig               45            205133   
16819                     Believer               23            204346   
11342                     Believer               23            203782   
7873                           Bod               55            192999   
18681                          Bod               55            192999   
1864                  Crazy in Lov               65            249106   
1383                  Crazy in Lov               65            227500   
824           Cum on Feel the Noiz               68            290960   
542           Cum on Feel the Noiz               68            287400   
7174                   Love No Mor               68            168635   
11879                  Love No Mor               68            168635   
6086                Pretty in Pink               53            239960   
2871                Pretty in Pink               53            238106   
18117                   Promiscuou               67            242200   
8606                    Promiscuou               67            242373   
10835            Set Fire to the R               75            242973   
15795            Set Fire to the R               75            241693   
12991             Steady, As She G               64            215266   
49                Steady, As She G               64            215266   

       acousticness  danceability  energy  instrumentalness  key  liveness  \
11960      0.299000         0.553   0.502          0.000000    0    0.7640   
5910       0.299000         0.541   0.522          0.000000    0    0.7750   
15407      0.747000         0.529   0.648          0.000000   11    0.1580   
4194       0.187000         0.582   0.668          0.000000   11    0.1140   
16819      0.051200         0.772   0.780          0.000000   10    0.1190   
11342      0.041700         0.772   0.775          0.000000   10    0.2260   
7873       0.301000         0.619   0.843          0.000000    2    0.0538   
18681      0.309000         0.620   0.841          0.000000    2    0.0538   
1864       0.000553         0.646   0.720          0.000008    7    0.0556   
1383       0.868000         0.606   0.326          0.000000    7    0.1200   
824        0.007190         0.409   0.920          0.001050   11    0.0823   
542        0.000183         0.441   0.908          0.003280   11    0.0612   
7174       0.020300         0.768   0.673          0.000045    0    0.2580   
11879      0.021600         0.767   0.674          0.000059    0    0.2570   
6086       0.117000         0.376   0.885          0.010400    2    0.2960   
2871       0.105000         0.386   0.916          0.009910    2    0.3250   
18117      0.050000         0.813   0.977          0.000028   10    0.1470   
8606       0.053600         0.807   0.970          0.000066   10    0.1490   
10835      0.004080         0.603   0.670          0.000002    2    0.1120   
15795      0.003820         0.610   0.683          0.000002    2    0.1250   
12991      0.008650         0.534   0.605          0.008260    9    0.0781   
49         0.013200         0.524   0.578          0.009210    9    0.1080   

       loudness audio_mode  speechiness    tempo time_signature  audio_valence  
11960    -5.972          1       0.0409  129.976              4          0.296  
5910     -5.997          1       0.0440  129.853              4          0.255  
15407    -7.895          0       0.0399  159.036              4          0.808  
4194     -8.585          0       0.0433  145.271              4          0.864  
16819    -4.288          0       0.1080  124.988              4          0.719  
11342    -4.388          0       0.1120  124.978              4 

Para tratar esses dados, decidi usar a média dos valores discrepantes.

In [117]:
#completando a lista de floats
ftrs_float.append("audio_valence")

#tirar a media dos discrepantes
summaries = {'song_duration_ms': 'mean'}
for ftr in ftrs_float:
    summaries[ftr] = "mean"

df = df.groupby(colunas).agg(summaries).reset_index()

#Verificar se deu tudo certo 
duplicates = df.duplicated(subset = colunas, keep=False)
df[duplicates].sort_values(by = "song_name")

Empty DataFrame
Columns: [song_name, song_popularity, key, audio_mode, time_signature, song_duration_ms, acousticness, danceability, energy, instrumentalness, liveness, loudness, speechiness, tempo, audio_valence]
Index: []

In [119]:
depois2 = len(df)
print(antes - depois2)

3914


Foram retirados ao todo 3914 duplicados.

In [123]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14901 entries, 0 to 14900
Data columns (total 15 columns):
song_name           14901 non-null object
song_popularity     14901 non-null int64
key                 14901 non-null int64
audio_mode          14901 non-null category
time_signature      14901 non-null category
song_duration_ms    14901 non-null float64
acousticness        14901 non-null float64
danceability        14901 non-null float64
energy              14901 non-null float64
instrumentalness    14901 non-null float64
liveness            14901 non-null float64
loudness            14901 non-null float64
speechiness         14901 non-null float64
tempo               14901 non-null float64
audio_valence       14901 non-null float64
dtypes: category(2), float64(10), int64(2), object(1)
memory usage: 1.5+ MB


## Novas colunas